### PR TITLE
[WIP] Example of DataPipes and DataFrames integration

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59053 [WIP] DataPipes Developer docs
* **#58597 [WIP] Example of DataPipes and DataFrames integration**
* #58153 [WIP] Add notebook with DataPipes API examples
* #58152 [WIP] Add simple `fork` DataPipe
* #58151 [WIP] Decorate ZipDataPipe as `zip`
* #58150 [WIP] Add simple `groupby` DataPipe
* #58149 [WIP] Make `batch` DataPipe sensitive to batch_level argument
* #58148 [WIP] Add simple `unbatch` DataPipe
* #58147 [WIP] Make `filter` DataPipe sensitive to batch_level argument
* #58146 [WIP] Make `shuffle`  DataPipe sensitive to batch_level argument
* #58145 [DataLoader] Rename `transform` DataPipe to `legacy_transform`

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
https://github.com/pytorch/pytorch/blob/d56f25977d3f0da078bec0dab01fb90861b73f28/torch/utils/data/dataframes_pipes.ipynb - notebook

CC @dzhulgakov @ejguan

File: torch/utils/data/dataframes_pipes.ipynb [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+   ],
+   "source": [
+    "def get_dataframes_pipe(range = 20, batch_size = 7):\n",
+    "    return ExampleIterPipe(range = range).map(lambda i: (i, i % 3)).to_dataframes_pipe(columns = ['i','j'], batch_size = batch_size)\n",
+    "\n",
+    "dp = get_dataframes_pipe()\n",
+    "\n",
+    "# Iterator converts DF Pipe into singular items\n",
+    "for i in dp:\n",
+    "    print(i)"
```
Comment (ejguan) [inline torch/utils/data/dataframes_pipes.ipynb:83]: I have same feeling with Francisco that DFPipe should represent one DF and yield rows.

File: torch/utils/data/dataframes_pipes.ipynb [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    "dp['y'] = dp.i * 100 + dp.j - 2.7\n",
+    "dp = dp.shuffle()\n",
+    "dp = dp - 17\n",
+    "dp['y'] = dp.y * 10000\n",
+    "dp = dp.batch(2)\n",
+    "for i,v in enumerate(dp):\n",
+    "    print('batch',i)\n",
+    "    for x in v:\n",
+    "        print(x)"
+   ]
```
Comment (ejguan) [inline torch/utils/data/dataframes_pipes.ipynb:231]: For `batch` over DFPipe, we can take a look here https://github.com/facebookexternal/torchdata/blob/main/torcharrow/torcharrow-batching.ipynb
to yield `df[0: batch_size], df[batch_size:batch_size*2], ...`

We can use `batch_size` as an internal state for DFPipe without actually split them into multiple DFs. So, we can still run ops over the whole DF. (It's kind like push batching logic to the end of the DF pipeline)

In this case, `batch` op over DFPipe won't change DFPipe object into regular DataPipe.

File: torch/utils/data/dataframes_pipes.ipynb [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    "for i in dp:\n",
+    "    print(i)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": []
```
Comment (ejguan) [inline torch/utils/data/dataframes_pipes.ipynb:299]: Another thing I want to mention is do we want to support `groupby` for DF?

So, the following DataPipe ops becomes two categories in this case:
- DFPipe remains after applied with the op
  - batch
  - map
  - concat ?
- DFPipe becomes DataPipe after applied with the op
  - shuffle
  - groupby

**Comment by VitalyFedyunin** [comment]
New version of the notebook https://github.com/pytorch/pytorch/blob/be4bd36a9c979b29846975e1386d96151ca8f6d7/torch/utils/data/dataframes_pipes.ipynb

**Comment by dzhulgakov** [comment]
The updated behavior with chunks looks really nice! A few questions:

> Chunk may have any number of DataFrames as soon as total number of rows equal to batch size.
Why do we need this complexity? Can we say that DataFrame itself is like a 1-level nested "chunk"? Then `df.batch().iterator()` should just return simple dataframes, because they if iterated upon would lead to rows. Just as a raw datapipe after batching returns you list/DataChunks. This behavior (up to level 1 of nesting) is also easier to replicate on Koski.

Do we need DataChunk as datastructure in general? Is it just to handle the nesting level in implementation? Maybe we can carry nesting level as a separate metadata on the pipe instead? I think utilizing raw lists would be less surprising to users (especially if one does `.batch().map()`)

**Comment by VitalyFedyunin** [comment]
Chunks removed https://github.com/pytorch/pytorch/blob/3170068d9c7e255ac1e3d59cc225f7b7e751d89f/torch/utils/data/dataframes_pipes.ipynb

**Comment by ejguan** [comment]
Seems really great. Just one thing in my mind:
> unbatch of list with DataFrame works similarly to regular unbatch. Note: DataFrame sizes might change

If DF sizes might change, how can we determine the size. Should we keep an internal state of `dataframe_size`? So, when no batching exists, we can revert DF back to `dataframe_size`.

File: torch/utils/data/dataframes_pipes.ipynb [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+   "source": [
+    "dp = get_dataframes_pipe(dataframe_size = 3)\n",
+    "dp['y'] = dp.i * 100 + dp.j - 2.7\n",
+    "for i in dp.raw_iterator():\n",
+    "    print(i)"
+   ]
+  },
+  {
+   "source": [
+    "`shuffle` of DataFramePipe effects rows in individual manner"
```
Comment (wenleix) [inline torch/utils/data/dataframes_pipes.ipynb:179]: Does that mean `shuffle` is not local to each DataFrame? (from the code looks so :) )

File: torch/utils/data/datapipes/iter/dataframes.py [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+
+    def __iter__(self):
+        for item in self.source_datapipe:
+            yield self.output_var.calculate_me(item)
+
+
+
+
+@functional_datapipe('dataframes_per_row', is_df = True)
+class ShuffleDataFramesPipe(DFIterDataPipe):
```
Comment (wenleix) [inline torch/utils/data/datapipes/iter/dataframes.py:22]: curious: why all the class name is `ShuffleDataFramesPipe`?

File: torch/utils/data/dataframes_pipes.ipynb [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+     "output_type": "stream",
+     "name": "stdout",
+     "text": [
+      "[       i      j\n0  1.111  1.111\n1  1.112  1.112\n2  1.113  1.113\n0  1.114  1.111\n1  1.115  1.112]\n[       i      j\n2  1.116  1.113\n0  1.117  1.111\n1  1.118  1.112\n2  1.119  1.113\n0  1.120  1.111]\n[(1.111, 0), (1.112, 1), (1.113, 2), (1.114, 0), (1.115, 1)]\n[(1.116, 2), (1.117, 0), (1.118, 1), (1.119, 2), (1.12, 0)]\n"
+     ]
+    }
+   ],
+   "source": [
+    "dp = get_dataframes_pipe(range = 10, dataframe_size = 3)\n",
+    "dp = dp.map(lambda x: x + 1111)\n",
```
Comment (wenleix) [inline torch/utils/data/dataframes_pipes.ipynb:383]: I probably miss something, but how is `map` over dataframe implemented? only found `dataframes_filter` but not for `map`.

**Review by wenleix — COMMENTED** [review]
Looks great.

**Comment by VitalyFedyunin** [comment]
UPD: Changed iterators as per discussion https://github.com/pytorch/pytorch/blob/f0fbfe1b41b2d368e38db1c6e2618b434df04fee/torch/utils/data/dataframes_pipes.ipynb

**Comment by dzhulgakov** [comment]
This looks awesome! Full thumbs up on API look and feel from my side :)

The only think is DataChunk - I think you need it only because you're using Pandas and its indexing is a bit weird. Following TA prototype, it does behave as a list, so `list(df)` would have the same behavior on DataFrame as on DataChunk. In that case I don't think you need DataChunk at all and can just return torcharrow.DataFrame in the batched case (as it does implement the superset of API). What do you think? If that's the case, probably add a TODO that DataChunk will be removed when proper dataframes appear so that we don't forget

**Comment by VitalyFedyunin** [comment]
> This looks awesome! Full thumbs up on API look and feel from my side :)
> 
> The only think is DataChunk - I think you need it only because you're using Pandas and its indexing is a bit weird. Following TA prototype, it does behave as a list, so `list(df)` would have the same behavior on DataFrame as on DataChunk. In that case I don't think you need DataChunk at all and can just return torcharrow.DataFrame in the batched case (as it does implement the superset of API). What do you think? If that's the case, probably add a TODO that DataChunk will be removed when proper dataframes appear so that we don't forget

I will add TODO to review and potentially nuke DataChunk as we migrate from pandas to TorchArrow
